source/extensions/dynamic_modules/sdk/rust/src/buffer.rs

```diff
@@ -49,3 +51,163 @@ impl EnvoyBuffer<'_> {
     unsafe { std::slice::from_raw_parts_mut(self.raw_ptr as *mut u8, self.length) }
   }
 }
+
+/// This implements the [`std::io::Read`] for reading request body data.
+pub struct RequestBodyReader<'a> {
+  raw_ptr: crate::abi::envoy_dynamic_module_type_http_filter_envoy_ptr,
+  _marker: std::marker::PhantomData<&'a ()>,
+  offset: usize,
+}
+
+impl std::io::Read for RequestBodyReader<'_> {
+  fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+    let mut size: usize = 0;
+    let ok = unsafe {
+      envoy_dynamic_module_callback_http_read_request_body(
+        self.raw_ptr,
+        self.offset,
+        buf.as_ptr() as _,
+        buf.len(),
+        &mut size,
+      )
+    };
+    if !ok {
+      return Result::Err(std::io::Error::new(
+        std::io::ErrorKind::Other,
+        "Request body not available",
+      ));
+    }
+    self.offset += size;
+    Result::Ok(size)
+  }
+}
+
+impl RequestBodyReader<'_> {
+  pub fn new(raw_ptr: crate::abi::envoy_dynamic_module_type_http_filter_envoy_ptr) -> Self {
+    Self {
+      raw_ptr,
+      _marker: std::marker::PhantomData,
+      offset: 0,
+    }
+  }
+}
+
+/// This implements the [`std::io::Write`] for writing request body data.
+pub struct RequestBodyWriter<'a> {
+  raw_ptr: envoy_dynamic_module_type_http_filter_envoy_ptr,
+  _marker: std::marker::PhantomData<&'a ()>,
+}
+
+impl RequestBodyWriter<'_> {
+  pub fn new(raw_ptr: envoy_dynamic_module_type_http_filter_envoy_ptr) -> Self {
+    Self {
+      raw_ptr,
+      _marker: std::marker::PhantomData,
+    }
+  }
+}
+
+impl std::io::Write for RequestBodyWriter<'_> {
+  fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+    let mut size: usize = 0;
+    let ok = unsafe {
+      envoy_dynamic_module_callback_http_write_request_body(
+        self.raw_ptr,
+        buf.as_ptr() as _,
+        buf.len(),
+        &mut size,
+      )
+    };
+    if !ok {
+      return Result::Err(std::io::Error::new(
+        std::io::ErrorKind::Other,
+        "Request body not available",
+      ));
+    }
+    Result::Ok(size)
+  }
+
+  fn flush(&mut self) -> std::io::Result<()> {
+    Result::Ok(())
+  }
+}
+
+/// This implements the [`std::io::Read`] for reading response body data.
+pub struct ResponseBodyReader<'a> {
```

Comment:
The request/response structs/impls are basically the same. Can they be shared somehow?